### PR TITLE
NFT and Registry modules

### PR DESCRIPTION
**Pull request by jaybutera**
Integrating NFT and VARegistry modules.
Supporting transfer, create_registry, and mint endpoints.

### CODE REVIEW

File: runtime/src/bridge.rs
Comment (mikiquantum) [inline runtime/src/bridge.rs:97]: Is there any reason why not naming this function `transfer_nft` ?

File: runtime/src/nft/mod.rs
Comment (mikiquantum) [inline runtime/src/nft/mod.rs:94]: Is this comment still relevant?

File: runtime/src/registry/mod.rs
Comment (mikiquantum) [inline runtime/src/registry/mod.rs:1]: Should we rename this Pallet, since `registry` is pretty generic. Maybe nft_registry?

File: runtime/src/registry/mod.rs
Comment (mikiquantum) [inline runtime/src/registry/mod.rs:92]: We should review all these weights before merging.

File: runtime/src/registry/mod.rs
Comment (mikiquantum) [inline runtime/src/registry/mod.rs:151]: So the registry id would be formed as:
```
id = [1,0,0,0,0,0,...]
nplus1 = U256(id).add(1) = [2,0,0,0,0,...]
```
so it would be a byte array nonce, right?

File: runtime/src/registry/mod.rs
Comment (mikiquantum) [inline runtime/src/registry/mod.rs:197]: Wouldn't this be None if registry_id doesnt exist? In which case the below check can be replace by is_ok_or(RegistryDoesNotExist).

File: runtime/src/registry/mod.rs
Comment (mikiquantum) [inline runtime/src/registry/mod.rs:211]: To be more precise, the registry property is the last one after the custom properties already defined in the registry. We currently have disabled the ability to pass more proofs, but if we always ensure that the registry_property is always the next one after the custom ones, we do not have to modify this logic if we add more proofs later on.

File: runtime/src/registry/mod.rs
Comment (mikiquantum) [inline runtime/src/registry/mod.rs:214]: Since you have stored the registry property as a required attribute in the registry, why dont compare as a whole?
`registry_prop == reg_info.fields[0]`

File: runtime/src/registry/types.rs
Comment (mikiquantum) [inline runtime/src/registry/types.rs:17]: Salt is Byte32 array, not sure if you want to constrain it here.

File: runtime/src/bridge.rs
Comment (mikiquantum) [inline runtime/src/bridge.rs:121]: As mentioned offline, I wanted to throw out there, if we needed to add a list of supported registries per chain_id, so there is more control on where to move the nft. Since as it stands, as soon as we whitelist another nft registry in ethereum, we would be able to "accidentally" use the resource_id of that registry instead of the correspondent one.
I believe that might be duty of the resource_id mapping that the bridge is responsible for, but lets put some thought into it.

**Review by mikiquantum — COMMENTED** [review]
Overall is great, clean, following composibility and flexible to iterate from it. Some clarification comments only.
Great Job @jaybutera

File: runtime/src/bridge.rs
Comment (jaybutera) [reply runtime/src/bridge.rs:97]: Asset over nft is the naming convention throughout the project. Inspired from Dan Forbes' naming convention to try to be more general. That being said I have no problem in changing this, let me know.

File: runtime/src/registry/mod.rs
Comment (jaybutera) [reply runtime/src/registry/mod.rs:1]: How about va_registry?

File: runtime/src/registry/mod.rs
Comment (jaybutera) [reply runtime/src/registry/mod.rs:151]: That's correct. The RegistryId is still an H160. I wanted to make a U160 type for efficiency and ease of use, but had version conflicts with Substrate when importing the shared crate, uint. May take a while to resolve and probably isn't critical.

File: runtime/src/registry/mod.rs
Comment (jaybutera) [reply runtime/src/registry/mod.rs:197]: I've been meaning to look deeper into this. For some reason, the StorageMap interface does not provide a try_get() - only get - [frame_support/storage/trait.StorageMap](https://substrate.dev/rustdocs/v2.0.0-rc6/frame_support/storage/trait.StorageMap.html).

I believe if a value doesn't exist, the default is returned (which is why a value must impl Default). Though I'm having trouble finding proof beyond [this point in the knowledge base](https://substrate.dev/docs/en/knowledgebase/runtime/storage#default-values).

It may be faster to compare the result to the default value but `contains_key` feels safer until I know for sure about the default values.

File: runtime/src/registry/mod.rs
Comment (jaybutera) [reply runtime/src/registry/mod.rs:214]: Good point. Is there any reason to worry about the padding here like there is in the solidity version?

Is it always an exact match or should it be
```registry_prop[..28] == reg_info.fields[n]```

File: runtime/src/nft/mod.rs
Comment (jaybutera) [reply runtime/src/nft/mod.rs]: Nope, removed 👍

File: runtime/src/registry/mod.rs
Comment (jaybutera) [reply runtime/src/registry/mod.rs]: Updated 👍

File: runtime/src/registry/types.rs
Comment (jaybutera) [reply runtime/src/registry/types.rs]: Updated 👍

File: runtime/src/bridge.rs
Comment (jaybutera) [reply runtime/src/bridge.rs:121]: Ok from my understanding, we need a mapping `m: C -> E`
```
E = set of Tinlake contracts (on Ethereum)
C = set of registries on Centrifuge
```
Not every registry necessarily maps to a contract in `E`, and its possible that multiple registries map to the same contract, so `m` is not injective. But every Tinlake contract in `E` has a corresponding registry in `C` so `m` is surjective.

So a substrate StorageMap `ResourceId => ResourceId` to represent `m` would be sufficient.

We can put this into the centrifuge `src/bridge.rs` and allow access through the bridge admin account. I'm not sure how the proposal process works in the parent chainbridge-pallet and if that can be integrated to?

File: runtime/src/bridge.rs
Comment (mikiquantum) [inline runtime/src/bridge.rs:107]: This check is already done in the `transfer_nonfungible` function call below.

File: runtime/src/bridge.rs
Comment (mikiquantum) [reply runtime/src/bridge.rs:121]: Yes that is correct. TInlake is just a consumer of the NFTs, so we can simplify to:
```
E = set of nft registries on Ethereum
C = set of registries on Centrifuge
```
I believe that it should be duty of the bridge contracts/pallets admin'd by the relayers (chainbridge-pallet and erc721-handler) that is whitelisted by the admin operators. Since eventually, the bridge will keep track of where the assets are moving to/from, sounds like a first step towards it.

A Centrifuge Asset can only be minted in the centrifuge protocol, so lets call it the NFT source of truth, so any move will always originate in the cent chain first `C -> E`. Since we always lock the asset before sending, when the NFT comes back `E -> C` (if it does), a check can be performed to verify that. the token_id exists in the desired registry, and owned by the bridge address, but that sounds like the duty of the centrifuge receiver function in `src/bridge.rs`. 

So a whitelist storage map in both sides `centrifuge-chain` and `ethereum`, and that means that we would need to modify the chainbridge-substrate pallet and the chainbridge erc721-handler contract.
That would mean that we need a new deployment on ethereum, and therefore a cascade of changes around most of the layers, which is a bit unfortunate.
There are some alternative solutions, and I would like put more thought. Let sync offline about it.

File: runtime/src/lib.rs
Comment (mikiquantum) [inline runtime/src/lib.rs:819]: I would add here root or collective with 3/4 majority. See:
- https://github.com/centrifuge/centrifuge-chain/blob/master/runtime/src/lib.rs#L807
- https://github.com/centrifuge/centrifuge-chain/blob/master/runtime/src/rad_claims.rs#L174

File: runtime/src/lib.rs
Comment (mikiquantum) [inline runtime/src/lib.rs:887]: Please add the new pallets always at the end, otherwise it can screw up the metadata indices

File: runtime/src/nft/mod.rs
Comment (mikiquantum) [inline runtime/src/nft/mod.rs:128]: Does this check if the destination account exists as of has the existential deposit?

File: pallets/bridge-names/src/lib.rs
Comment (mikiquantum) [inline pallets/bridge-names/src/lib.rs:1]: Maybe rename to `bridge-mappings`

**Review by mikiquantum — APPROVED** [review]
Overall looks good!

File: runtime/src/nft/mod.rs
Comment (jaybutera) [reply runtime/src/nft/mod.rs]: `AccountId::default()` is the return value if no owner exists. The approach is confusing so I switched to using an Option type check and removed this line.

**Pull request merged**
